Zoom Recordings

jwalits opened the pull request:
Hi,

This pull request builds on the initial work done as part of #125 to support display of recordings in Moodle associated with the Zoom meeting.

Fixes #46
Supercedes #125

abias commented:
Hi @jrchamp ,

may I ask you to have a look on this PR as soon as you have time? Parts of the code have already been discussed in #125. Based on this code, ETH Zürich has funded the development to rebase and extende the code to automatically fetch the recordings from Zoom.

Looking forward to your reply.

Cheers,
Alex

jrchamp commented on locallib.php:
```diff
@@ -799,3 +810,56 @@ function zoom_get_alternative_host_array_from_string($alternativehoststring) {
+    global $DB;
+
+    $recordings = $DB->get_records('zoom_meeting_recordings', ['zoomid' => $zoomid]);
+    foreach ($recordings as $key => $recording) {
+        unset($recordings[$key]);
+        $recordings[$recording->zoomrecordingid] = $recording;
```
Probably better to build a fresh array than use unset.

jrchamp commented on locallib.php:
```diff
@@ -799,3 +810,56 @@ function zoom_get_alternative_host_array_from_string($alternativehoststring) {
+function zoom_get_meeting_recordings_grouped($zoomid) {
+    global $DB;
+
+    $records = $DB->get_records('zoom_meeting_recordings', ['zoomid' => $zoomid], 'timecreated ASC');
```
Wouldn't we care more about the recording start time?
```suggestion
    $records = $DB->get_records('zoom_meeting_recordings', ['zoomid' => $zoomid], 'recordingstart ASC');
```

jrchamp commented on loadrecording.php:
```diff
@@ -0,0 +1,67 @@
+list($course, $cm, $zoom) = zoom_get_instance_setup();
+
+if (!$rec = $DB->get_record('zoom_meeting_recordings', array('id' => $recordingid))) {
+    print_error('Recording could not be found');
```
From the Moodle codechecker:

> print_error() has been deprecated; use throw new moodle_exception()

jrchamp commented on showrecording.php:
```diff
@@ -0,0 +1,56 @@
+// Find the video recording and audio only recording pair that matches the criteria.
+$recordings = $DB->get_records('zoom_meeting_recordings', array('meetinguuid' => $meetinguuid, 'recordingstart' => $recordingstart));
+if (empty($recordings)) {
+    print_error('Recordings could not be found');
```
From the Moodle codechecker:

> print_error() has been deprecated; use throw new moodle_exception()

jrchamp commented on loadrecording.php:
```diff
@@ -0,0 +1,67 @@
+
+list($course, $cm, $zoom) = zoom_get_instance_setup();
+
+if (!$rec = $DB->get_record('zoom_meeting_recordings', array('id' => $recordingid))) {
```
Can we move this database query below the require_login and capability checks?

jrchamp commented on recordings.php:
```diff
@@ -0,0 +1,143 @@
+echo $OUTPUT->heading($strname);
+
+$iszoommanager = has_capability('mod/zoom:addinstance', $context);
+
```
This is another location where I wonder what the purpose of the configuration value `zoom/viewrecordings` is. It's difficult for me to tell if it's being implemented consistently.

jrchamp commented on showrecording.php:
```diff
@@ -0,0 +1,56 @@
+$recordingstart = required_param('recordingstart', PARAM_INT);
+
+// Find the video recording and audio only recording pair that matches the criteria.
+$recordings = $DB->get_records('zoom_meeting_recordings', array('meetinguuid' => $meetinguuid, 'recordingstart' => $recordingstart));
```
Can we move this query below the require_capability?

jrchamp commented on showrecording.php:
```diff
@@ -0,0 +1,56 @@
+
+// Toggle the showrecording value.
+foreach ($recordings as $rec) {
+    $rec->showrecording = intval($rec->showrecording) === 0 ? 1 : 0;
```
This toggle makes the toggle flip back any forth each time this page is visited. That's not [idempotent](https://developer.mozilla.org/en-US/docs/Glossary/Idempotent). Additionally, you want all GET requests to be "side effect" free. This is particularly important due to the risk of [replay attacks](https://en.wikipedia.org/wiki/Replay_attack).

jrchamp reviewed (commented):
There's a lot here, so I did as much as I could with the little energy that remains to me from our first week of classes.

jrchamp commented:
Also, there's a handful of codechecker warnings in the Actions output.

jwalits commented on classes/webservice.php:
```diff
@@ -761,4 +761,47 @@ public function encode_uuid($uuid) {
+     * @return array Returns the list of recording URLs and the type of recording that is being sent back.
+     */
+    public function get_recording_url_list($meetingid) {
+        $url = 'meetings/' . $meetingid . '/recordings';
+        $settingsurl = 'meetings/' . $meetingid . '/recordings/settings';
```
i saw another place in the code is using the `encode_uuid();` so using that here, instead of rawurlencode for consistency

jwalits commented:
Hi @jrchamp 

I have updated most of the concerns raised in your initial review. Could you please enable codechecker so I can see if I covered all the complaints raised by Github Actions.

Thanks

abias commented:
@jrchamp May I ping you about the request by @jwalits 10 days ago? As far as I understand, he needs a little action from you before he can proceed.

jrchamp commented:
> @jrchamp May I ping you about the request by @jwalits 10 days ago? As far as I understand, he needs a little action from you before he can proceed.

The codechecker runs automatically for users that have the Contributor tag, so that little action is already complete. I'll work on reviewing the changes soon.

jrchamp commented on showrecording.php:
```diff
@@ -0,0 +1,62 @@
+
+$now = time();
+
+// Toggle the showrecording value.
+if ($showrecording === 1 || $showrecording === 0) {
+    foreach ($recordings as $rec) {
+        $rec->showrecording = $showrecording;
+        $rec->timemodified = $now;
+        $DB->update_record('zoom_meeting_recordings', $rec);
+    }
+}
```
This is idempotent now, so thank you. My second concern is that it may allow for [Cross-site Request Forgery (CSRF)](https://cwe.mitre.org/data/definitions/352.html)

Here's some Moodle documentation that may help address the problem:
https://docs.moodle.org/dev/Security:Cross-site_request_forgery#What_you_need_to_do_in_your_code

jwalits replied:
I have added the confirm_sesskey code to the showrecording to address this issue.

jwalits commented on classes/task/get_meeting_recordings.php:
```diff
@@ -0,0 +1,118 @@
+                    foreach ($zoomrecordingpairlist as $recordingstarttime => $zoomrecordingpair) {
+                        // The video recording and audio only recordings are grouped together by their recording start timestamp.
+                        foreach ($zoomrecordingpair as $zoomrecordinginfo) {
+                            if (isset($recordings[trim($zoomrecordinginfo->meetinguuid)])) {
+                                mtrace('Recording(s) for meetinguuid: ' . $zoomrecordinginfo->meetinguuid . ' exist(s)...skipping');
+                                unset($recordings[trim($zoomrecordinginfo->meetinguuid)]);
+                                continue(2);
+                            }
```
@jrchamp @abias - I have been trying to replicate this, but came across a bigger issue, for which I'll need to update the way we handle deletion. We'll most likely have to handle deletion in a separate scheduled task. The issue is related to the way the API for `/recordings` works. https://marketplace.zoom.us/docs/api-reference/zoom-api/cloud-recording/recordingget
```
To get Cloud Recordings of a meeting, provide the meeting ID or meeting UUID.
If the meeting ID is provided instead of UUID,the response will be for the latest meeting instance. 

To get Cloud Recordings of a webinar, provide the webinar ID or the webinar UUID. If the webinar ID is provided instead of UUID,the response will be for the latest webinar instance. \n\nIf a UUID starts with \"/\" or contains \"//\" (example: \"/ajXp112QmuoKj4854875==\"), you must **double encode** the UUID before making an API request.
```

When you pass a meeting id to it, it will only return the recordings of the latest meeting instance. The way the deletion is setup, the code removes all meeting instances that aren't present in the response. This causes a problem where previous recordings from the same zoom meeting get deleted in Moodle.

I will need to update the way we handle this.

jwalits replied:
In the recent round of testing, I found that code is not quite correct. You can create multiple recordings for the same UUID (by stopping and restarting the recording in the same meeting). I have updated the code to handle that.

jwalits commented on loadrecording.php:
```diff
@@ -0,0 +1,66 @@
+ * Load zoom meeting recording and add a record of the view.
+ *
+ * @package    mod_zoom
+ * @copyright  2020 UC Regents
```
This code was added from the original PR, I didn't touch that part. If @abias has a preference on that I can update that, otherwise I'm ok to keep that as is.

abias replied:
My gut feeling is that, if the code was written by UC Regents during the original PR and left untouched during the recent refinements, the credits should be kept as they are.

jrchamp reviewed (commented):
It looks like something may have happened with a merge conflict as there's 95 commits that have changed. I highlighted a couple that seemed like they were leftover from other changes.

jrchamp commented on classes/task/delete_meeting_recordings.php:
```diff
@@ -0,0 +1,104 @@
+ * The task for deleting recordings in Moodle if removed from Zoom.
+ *
+ * @package    mod_zoom
+ * @author     Jwalit Shah <jwalitshah@catalyst-au.net>
```
Moodle PHPDoc requires a file-level `@copyright` here.

jrchamp commented on classes/task/delete_meeting_recordings.php:
```diff
@@ -0,0 +1,104 @@
+        return get_string('getmeetingrecordings', 'mod_zoom');
+    }
+
+    public function execute() {
```
Moodle PHPDoc requires a function-level documentation comment here similar to the one above for get_name().

jrchamp commented on classes/task/get_meeting_recordings.php:
```diff
@@ -0,0 +1,107 @@
+ * The task for getting recordings from Zoom to Moodle.
+ *
+ * @package    mod_zoom
+ * @author     Jwalit Shah <jwalitshah@catalyst-au.net>
```
Moodle PHPDoc requires a file-level `@copyright` here.

jrchamp commented on classes/task/get_meeting_recordings.php:
```diff
@@ -0,0 +1,107 @@
+        return get_string('getmeetingrecordings', 'mod_zoom');
+    }
+
+    public function execute() {
```
Moodle PHPDoc requires a function-level documentation comment here similar to the one above for get_name().

jrchamp reviewed (commented):
Here's the 4 PHPDoc errors that are signaling the GitHub Action failures.

jwalits commented:
Hi @jrchamp 

I've fixed up the missing PHPDocs but not quite sure of the why Actions is failing - appears to be failing on grunt? I haven't touched the javascript files for this PR (just rebased my branch off the master branch).

Do you think, I should rebuild the JS files locally and push new updates?

jrchamp commented:
@jwalits No, you don't need to modify the JavaScript files. The fix for that was merged in #325. I'll try flattening the code to get a clean rebase.

jrchamp commented on locallib.php:
```diff
@@ -534,6 +535,17 @@ function zoom_is_user_not_found_error($error) {
+ * @param moodle_exception $error
+ * @return bool
+ */
+function zoom_recording_not_found_error($error) {
```
I don't see where this new function is used: `zoom_recording_not_found_error()`

jrchamp reviewed (commented):
Hi @jwalits 

I consolidated the 30 commits into a single commit so that we would have a clean workspace, because some of the commits seemed to be from older, merged pull requests. I identified and removed a few changes that were due to merge conflicts and bumped the upgrade savepoint and plugin version to allow for easier testing.

Please take a look to make sure that all of the code is still here in this Pull Request. I noticed one new function is not used, so I wasn't completely confident that something may have been lost in the recent couple of rebase shuffles.

jwalits commented:
Hi @jrchamp 

Thanks for the update. I'll have another look through the code today and make sure all the recording code is all there and working correctly.

jwalits commented:
Hi @jrchamp 

I have updated the PR with an additional commit and rebased it after the latest changes. It appeared like some of the functions related to recordings got removed from locallib.php. I have re-added those, did another test locally and all appears to be working as expected. 

I have also removed the code that you flagged as being unused. I couldn't find any references in the code and in my testing, I couldn't see any need for that. It might've been some legacy code that came in from the originally submitted PR.

Could you please take another look and see if its looking good to merge it in?

jrchamp commented on recordings.php:
```diff
@@ -0,0 +1,147 @@
+require_once(__DIR__ . '/lib.php');
+require_once(__DIR__ . '/locallib.php');
+
+$id = required_param('id', PARAM_INT);
+
+$config = get_config('zoom');
+
```
Neither of these variables appear to be used.
```suggestion
```

jrchamp commented on loadrecording.php:
```diff
@@ -0,0 +1,66 @@
+require_once($CFG->libdir . '/moodlelib.php');
+require_once(__DIR__ . '/locallib.php');
+
+$id = required_param('id', PARAM_INT);
```
This variable is not used.
```suggestion
```

jwalits replied:
The id param gets used by zoom_get_instance_setup() method to get the relevant info
`list($course, $cm, $zoom) = zoom_get_instance_setup();`

jwalits replied:
Have removed that variable now from loadrecording.php

jrchamp commented on showrecording.php:
```diff
@@ -0,0 +1,65 @@
+require_once($CFG->libdir . '/moodlelib.php');
+require_once(__DIR__ . '/locallib.php');
+
+$id = required_param('id', PARAM_INT);
+$meetinguuid = required_param('meetinguuid', PARAM_TEXT);
+$recordingstart = required_param('recordingstart', PARAM_INT);
+$showrecording = required_param('showrecording', PARAM_INT);
+
+list($course, $cm, $zoom) = zoom_get_instance_setup();
+require_login($course, true, $cm);
+
+$context = context_module::instance($id);
+$PAGE->set_context($context);
+require_capability('mod/zoom:addinstance', $context);
+
+$urlparams = array('id' => $id);
```
As we have the course module information, it seems more consistent with the rest of the code in this PR to use that ID instead.
```suggestion
$meetinguuid = required_param('meetinguuid', PARAM_TEXT);
$recordingstart = required_param('recordingstart', PARAM_INT);
$showrecording = required_param('showrecording', PARAM_INT);

list($course, $cm, $zoom) = zoom_get_instance_setup();
require_login($course, true, $cm);

$context = context_module::instance($cm->id);
$PAGE->set_context($context);
require_capability('mod/zoom:addinstance', $context);

$urlparams = array('id' => $cm->id);
```

jwalits replied:
Removed the $id parameter from showrecording.php

jrchamp commented on loadrecording.php:
```diff
@@ -0,0 +1,66 @@
+
+require_capability('mod/zoom:view', $context);
+
+if (!$rec = $DB->get_record('zoom_meeting_recordings', array('id' => $recordingid))) {
```
Need to make sure that the use is allowed to view the recording:

1. Recording ID must be associated with the $zoom meeting. Can probably achieve this by adding `'zoomid' => $zoom->id` to the clause array.
2. Recording viewing must be allowed. Can probably achieve this by adding `'showrecording' => 1` to the clause array.

jwalits replied:
Thanks for the update on the missed threads from above @jrchamp 

I will have a look through those today and update the PR.

jwalits replied:
@jrchamp I have updated the code to address the points you raised. I have added a new activity level setting for checking whether new recordings are visible by default. Have also updated the handling around the zoom/viewrecordings config setting on the load/show/recordings pages.

In regards to your point:
> On the other end of the spectrum, a sneaky user is able to view any Zoom recording on the entire server via loadrecording.php, as long as they have view access for a Zoom meeting.

That is a valid concern, if we want to be stricter around that, we could possibly add a user token parameter (using the get_user_key method from core Moodle) when generating the loadrecording and validate it on that page. This way only valid users who clicked the link from the recordings page will be able to load the recording?

In regards to issues related to fetching/deleting recordings, most of them are resolved now, as we had to move the delete functionality to a new scheduled task: delete_meeting_recordings - but happy to reconsider any points that may still be outstanding and of concern.

jrchamp reviewed (commented):
I'm looking at the "files changed" view https://github.com/ncstate-delta/moodle-mod_zoom/pull/292/files and there's a couple design decision questions that are unresolved, such as:

- What is the single purpose of the `$config->viewrecordings` setting? It appears to be used to show and hide a recordings section on the view page, but in another context, it is used as the default visibility setting for newly created recordings. So for a user to know they exist at all, it seems to require that all recordings default to visible? That seems like the configuration value is being used for two separate purposes and even then, not consistently. If "show recordings in Zoom" is off at the system level, then the task probably should not be running (it would be a waste of resources if it's never going to be used). The linked pages (recordings.php, showrecording.php, loadrecording.php) should also probably not do anything when recordings are "off". The second purpose that the setting is being used for is "are recordings visible by default?". That should probably be a separate setting, ideally an activity default that instructors can override.
- On the other end of the spectrum, a sneaky user is able to view any Zoom recording on the entire server via loadrecording.php, as long as they have view access for a Zoom meeting.
- `zoom_get_all_meeting_records()` is still using meeting IDs that will expire as the array keys and is not checking the exists_on_zoom value. There's a series of comments on that issue from August that are still visible and relevant.

jrchamp commented on classes/task/get_meeting_recordings.php:
```diff
@@ -0,0 +1,113 @@
+                    get_string('strftimedaydatetime', 'core_langconfig')));
+            return;
+        }
+
```
We should probably add a check here for `$config->viewrecordings` so that this task does not run unless it is needed.

jwalits replied:
Sure, have added a check for viewrecordings before executing the task.

jrchamp commented on showrecording.php:
```diff
@@ -0,0 +1,68 @@
+list($course, $cm, $zoom) = zoom_get_instance_setup();
+require_login($course, true, $cm);
+
+$context = context_module::instance($id);
```
```suggestion
$context = context_module::instance($cm->id);
```

jwalits replied:
updated

jrchamp commented on db/upgrade.php:
```diff
@@ -635,5 +635,69 @@ function xmldb_zoom_upgrade($oldversion) {
+            $dbman->create_table($table);
+        }
+
+        // Add new field for recording_visible_default.
+        $table = new xmldb_table('zoom');
+        // Define field recording_visible_default to be added to zoom.
+        $field = new xmldb_field('recordings_visible_default', XMLDB_TYPE_INTEGER, '1', null, XMLDB_NOTNULL, null, '1',
+                'alternative_hosts');
+
+        // Conditionally launch add field recording_visible_default.
```
```suggestion
        // Add new field for recordings_visible_default.
        $table = new xmldb_table('zoom');
        // Define field recordings_visible_default to be added to zoom.
        $field = new xmldb_field('recordings_visible_default', XMLDB_TYPE_INTEGER, '1', null, XMLDB_NOTNULL, null, '1',
                'alternative_hosts');

        // Conditionally launch add field recordings_visible_default.
```

jwalits replied:
Updated the typo

jrchamp commented on mod_form.php:
```diff
@@ -516,6 +516,13 @@ public function definition() {
 
+        // Adding option for Recording Visiblity by default.
+        $mform->addElement('header', 'general', get_string('recording', 'mod_zoom'));
+        $mform->addElement('advcheckbox', 'recording_visible_default', get_string('recordingvisibility', 'mod_zoom'),
+                get_string('yes'));
+        $mform->setDefault('recording_visible_default', 1);
+        $mform->addHelpButton('recording_visible_default', 'recordingvisibility', 'mod_zoom');
```
```suggestion
        $mform->addElement('advcheckbox', 'recordings_visible_default', get_string('recordingvisibility', 'mod_zoom'),
                get_string('yes'));
        $mform->setDefault('recordings_visible_default', 1);
        $mform->addHelpButton('recordings_visible_default', 'recordingvisibility', 'mod_zoom');
```

jwalits replied:
Thanks for picking up the typo, have updated the code

jrchamp commented on recordings.php:
```diff
@@ -0,0 +1,147 @@
+$service = new mod_zoom_webservice();
+
+$now = time();
+if ($zoom->recurring || $now > (intval($zoom->start_time) + intval($zoom->duration))) {
```
Can this condition be removed? The instructor might be editing and re-using an activity, so if it has visible recordings, let's just show them.

jrchamp commented on recordings.php:
```diff
@@ -0,0 +1,147 @@
+            $table->data[] = [$recordingdate, $recordinghtml, $recordingpasscode, $recordingshowhtml];
+        }
+    }
+} else {
+    $cell = new html_table_cell();
+    $cell->colspan = count($table->head);
+    $cell->text = get_string('norecordings', 'mod_zoom');
+    $cell->style = 'text-align: center';
+    $row = new html_table_row([$cell]);
+    $table->data = [$row];
+}
```
This can be removed if the matching "if" condition is removed.

jwalits replied:
No issues, have remove the if and matching else condition

jrchamp commented on settings.php:
```diff
@@ -151,6 +151,11 @@
 
+    $viewrecordings = new admin_setting_configcheckbox('zoom/viewrecordings',
+            get_string('option_view_recordings', 'mod_zoom'),
+            '', 1, 1, 0);
```
Not all users have Cloud Recording available, so I feel like it may be better to default this to "off". Thoughts?

jwalits replied:
Sure, no issues from my end having it switched off by default.

jrchamp reviewed (commented):
Thanks @jwalits for getting those changes made so quickly.

I noticed a few small things, but you did a great job of addressing the concerns I mentioned before.

jrchamp commented on loadrecording.php:
```diff
@@ -0,0 +1,75 @@
+
+// Keep track of whether someone has viewed the recording or not.
+if ($view = $DB->get_record('zoom_meeting_recordings_view', $params)) {
+    if (!$view->viewed) {
```
```suggestion
    if (empty($view->viewed)) {
```

jrchamp commented on classes/task/get_meeting_recordings.php:
```diff
@@ -0,0 +1,116 @@
+        } else if (empty($config->apisecret)) {
+            mtrace('Skipping task - ', get_string('zoomerr_apisecret_missing', 'zoom'));
+            return;
+        } else if (!$config->viewrecordings) {
```
```suggestion
        } else if (empty($config->viewrecordings)) {
```

jrchamp commented on loadrecording.php:
```diff
@@ -0,0 +1,75 @@
+    'showrecording' => 1,
+    'zoomid' => $zoom->id,
+];
+if (!$rec = $DB->get_record('zoom_meeting_recordings', $params)) {
```
```suggestion
$rec = $DB->get_record('zoom_meeting_recordings', $params);
if (empty($rec)) {
```

jrchamp commented on locallib.php:
```diff
@@ -1199,3 +1199,61 @@ function zoom_sync_meeting_tracking_fields($zoomid, $trackingfields) {
+    global $DB;
+
+    $params = [];
+    if ($zoomid != null) {
```
```suggestion
    if ($zoomid !== null) {
```

jrchamp commented on view.php:
```diff
@@ -253,6 +253,18 @@
 }
 
+// Show recordings section if option enabled to view recordings.
+if ($config->viewrecordings) {
```
```suggestion
if (!empty($config->viewrecordings)) {
```

jrchamp commented on classes/task/delete_meeting_recordings.php:
```diff
@@ -0,0 +1,110 @@
+        $zoomrecordings = zoom_get_meeting_recordings();
+        // Fetch all recordings for the unique meetinguuid.
+        $meetinguuidsfetched = [];
+        if ($zoomrecordings) {
```
```suggestion
        if (!empty($zoomrecordings)) {
```

jrchamp reviewed (commented):
This is looking really great. I realized it would be good to only show the activity-level setting when the site-level viewrecordings setting is "on". The rest of the comments are tiny suggestions that caught my eye, like strict comparisons, avoiding assignments in conditionals, and avoiding implicit bool conversion.

jwalits commented:
Thanks @jrchamp for flagging those. I have updated the code with those

abias commented:
Hey @jrchamp ,

thank you for your thorough code review and for taking the time to review this larger big contribution at all.
I am grateful to see that there is still progress in this PR, but I am also sad to see that we haven't reached the goal yet. 

I would like to highlight that ETH Zürich who is funding this development is running out of time and the deployment for the next lecture term should have happened already two weeks ago.

May I ask you to have a look at @jwalits 's latest fixes and to run a last test so that this contribution can then hopefully be merged and released?

Thanks in advance,
Alex

jrchamp commented:
Hi @abias, I'm with you - it would be great to have this complete and merged. I submitted this for testing before I went on vacation, but I think everyone else was on vacation too (Thursday, November 25 and Friday, November 26 were university holidays for us).

abias commented:
Thank you, @jrchamp , for your reply and for pushing this now to the testing queue.

sgrandh3 commented:
Performed the following tests successfully:
Turned the view recordings setting on , and recorded both on the computer and on the cloud. Was able to retrieve them.
Performed only audio, audio and video recordings.
Was not able to test for the expired sessions (as it has to be 30days old)

sgrandh3 commented:
A small label change in the settings: The task name should be "Delete meeting recordings from Zoom" for \mod_zoom\task\delete_meeting_recordings
![image](https://user-images.githubusercontent.com/90855917/144061825-4baf569f-c289-4d98-9259-9ab6f7e3aefd.png)

jrchamp commented on classes/task/delete_meeting_recordings.php:
```diff
@@ -0,0 +1,110 @@
+     * @return string
+     */
+    public function get_name() {
+        return get_string('getmeetingrecordings', 'mod_zoom');
```
As @sgrandh3 identified, this should probably be:
```suggestion
        return get_string('deletemeetingrecordings', 'mod_zoom');
```
and a new string should be added for `deletemeetingrecordings` in lang/en/zoom.php

jwalits replied:
Thank you for the testing comments @sgrandh3 and the note about incorrect label. I have updated the lang string to resolve this issue.

sgrandh3 commented:
I have recorded meetings for 2 different zoom meetings say 1 and 2, on deleting meeting 1 (with recordings), i am not able to view other recorded meetings in the meeting  2.

jrchamp reviewed (approved):
Marking approved. Merge imminent.

abias commented:
Thank you very much for your patience and effort with this PR!
I am really grateful that it has been finally merged :)

jwalits commented:
Thanks @jrchamp and @sgrandh3 for the wonderful feedback during the review process and merging the PR. Thanks @abias for all your patience through this process.
